addons/crm/models/crm_lead.py

```diff
@@ -1491,6 +1486,76 @@ def _pls_increment_frequency(self, frequencies, field, value, won, lost):
         return frequencies
 
+    def _pls_reach_state(self, state):
+        """
+        When reaching won (or lost) state, by set won or set lost e.g.,
+        we need to increment the lead parameters in won (or lost) count in the frequency table,
+        to ensure the next probability computation will be based on up to date data.
+        """
+        self._pls_update_frequencies(is_won=state == 'won')
+
+    def _pls_leave_state(self, state):
+        """
+        When leaving a won (or lost) state, when restoring a lead e.g.,
+        we need to decrement the lead parameters in won (or lost) count in the frequency table,
+        to ensure the next probability computation will be based on up to date data.
+        """
+        self._pls_update_frequencies(is_won=state == 'won', decrement=True)
+
+    def _pls_update_frequencies(self, is_won=True, decrement=False):
+        """
+        When losing or winning a lead, this method is called to increment each PLS parameter related to the lead
+        in won_count (if won) or in lost_count (if lost).
+
+        This method is also used when reactivating a mistakenly lost lead (using the decrement argument).
+        In this case, the lost count should be de-increment by 1 for each PLS parameter linked ot the lead.
+        """
+        pls_start_date = self._pls_get_safe_start_date()
+        if not pls_start_date:
+            return
+        # Only treat leads created after the PLS start Date
+        pls_leads = self.filtered(
+            lambda lead: fields.Date.to_date(pls_start_date) >= fields.Date.to_date(lead.create_date))
+
+        Frequency = self.env['crm.lead.scoring.frequency'].sudo()
+        count_field_to_update = 'won_count' if is_won else 'lost_count'
+        anti_count_field_to_create = 'lost_count' if is_won else 'won_count'
+        increment_value = 1 if not decrement else -1
+
+        leads_pls_values = pls_leads._pls_get_lead_pls_values()
+
+        for lead in pls_leads:
+            frequencies_to_create = []
+            lead_pls_values = leads_pls_values[lead.id]
+            for field, value in lead_pls_values['values']:
+                if field == 'team_id':  # ignore team_id as frequencies are split by team_id.
+                    continue
+
+                frequency = Frequency.search([
+                    ('variable', '=', field),
+                    ('value', '=', value),
+                    ('team_id', '=', lead.team_id.id)
+                ], limit=1)  # should only return max 1 line anyway.
+                if frequency:
+                    frequency.write({
+                        count_field_to_update: frequency[count_field_to_update] + increment_value
+                    })
+                else:
+                    values_to_create = {
+                        'variable': field,
+                        'value': value,
+                        count_field_to_update: increment_value + 0.1,
```

Comment:
Maby split _pls_update_frequency_table_fields(_table) and try to reuse the same code.